### ISSUE TITLE
Latex verbatim environment does not wrap text

### ISSUE THREAD

**Issue by t-makaro** [t-makaro/nb_pdf_template#3]
Long lines of text will simply flow off the right side of the page.
Classic.tplx spills over at 86+1
Classicm.tplx spills over at 94+1
On the bright side, latex uses monospace fonts so this will be consistent.

I've found 3 potential solutions:
1. fancyvrb + listings
2. jinja wordwrap
3. [Sphinx's solution](https://github.com/sphinx-doc/sphinx/pull/2343)

**Comment by t-makaro** [comment]
fancyvrb + listings solutions is to add:
```
\usepackage{listings}
\lstset{ 
    breaklines=true,        % sets automatic line breaking
    breakatwhitespace=true,   
    fancyvrb=true,
    morefvcmdparams=\PY 2,
    postbreak=\mbox{\textcolor{red}{$\hookrightarrow$}\space},
}
```
Pros:
1. Pure LaTeX solution

Cons:
1. Doesn't play well with syntax highlighting. It would need to be turned off for code cells.
2. Messes with monospacing

**Comment by t-makaro** [comment]
jinja wordwrap solution:
Jinja has a [built-in filter](http://jinja.pocoo.org/docs/2.10/templates/#wordwrap) for word wrapping.

Pros:
1. Less of a bodge than making fancyvrb and listings work together
2. Doesn't mess with monospacing

Cons:
1. If placed before syntax highlighting, highlighting may not work properly. If placed after syntax highlighting, wordwrap will appear to wrap a random number of characters. It would need to be turned off for code cells.

**Comment by t-makaro** [comment]
Sphinx's Solution:
I've not looked into this one in too much detail. I will follow it up later.

**Comment by t-makaro** [comment]
Ultimately, line wrapping of outputs is far more important than code cells. The author of the notebook, can easily control the length of the lines of the code (pep8 compliance and you won't have an issue), but the output is more difficult (without having unnecessary for loops or something).

**Comment by vv111y** [comment]
Hi, I'm dealing with this issue right now. I thought since Sphinx had done it, it would be in nbsphinx.  That didn't work. I'm looking for a solution right now.  Your link for the jinja solution is broken. Could you direct me to it?  Also, any idea if this can be applied somewhere in nbsphinx templates?  Thanks.

**Comment by t-makaro** [comment]
I believe that I've fixed my link. I hadn't heard of nbsphinx until now, but it appears that they have hard coded the jinja template with only the ability to add to the start/end. I've really only got experience with nbconvert templates though. I'm planning to come back to this issue when I have time in a couple weeks.

**Comment by vv111y** [comment]
Link works, thanks.  
This needs to be dealt with so I'll follow along once you start.  I'll see what I can do as well.

**Comment by t-makaro** [comment]
I started a branch to try using the jinja wordwrap filter method, but I'm encountering [this issue](https://github.com/pallets/jinja/issues/175).

Wordwrap works great on single blocks of text, but it absolutely butchers things like numpy arrays.

**Comment by t-makaro** [comment]
@vv111y If you have any feedback for my jinja2 filtering solution, I'd appreciate it. I'd like to look at sphinx's solution, but I believe that this solution so much cleaner. Sphinx's solution could be added later to add text wrapping to code, but right now I'm happy to package in the wordwrap solution and ship version 1.2.

**Comment by mgeier** [comment]
Hi, I'm the author of `nbsphinx`. I've stumbled upon this project today, it looks great, and it inspired me to improve the LaTeX output of `nbsphinx` in https://github.com/spatialaudio/nbsphinx/pull/218. A temporary example PDF file is there: https://readthedocs.org/projects/nbsphinx/downloads/pdf/latex-codecells/ (once the PR is merged this will be in https://readthedocs.org/projects/nbsphinx/downloads/pdf/latest/).

With this new PR `nbsphinx` will be using the line-break and page-break features of Sphinx.

If you have suggestions for improvements, I'd love to hear them!

**Comment by t-makaro** [comment]
That looks fantastic!

It looks like you're doing a lot of work in the ```\newenvironment{nbsphinxfancyoutput}```. I opted to use ```\usepackage[breakable]{tcolorbox}``` to wrap the ```fancyvrb``` verbatim environment since ```tcolorbox```, and ```fancyvrb``` both page break nicely. I had issues with using ```minipage``` since I couldn't get it to page break. I'm also not build on sphinx so I don't have their options. But, since sphinx has done the hard work of line/page breaking ```fancyvrb```, you might be able to use ```tcolorbox``` in a similar manner to what I've done.

I'm curious how did you manage to make the prompts appear to the left of the cell? Presumably something to do with the use of ```"minipage```? My solution works, but it's a little hacky, so I prefer to put the ```In/Out [ ]``` prompts above the cell which also allows smaller margins without pushing code or prompts off the side of the page.

You should test with h5 and h6 markdown headers directly before a code cell. These get translated into ```\paragraph``` and ```\subparagraph``` in LaTeX. These don't line break before hand. It was causing me some issues.

**What really interests me is line breaking.** Currently, for output I just wrap the text at a known limit which works great because it's not highlighted. For code, I use Latex's minted package with relies on a shell escape. It's way slower and annoying to set up. I'd love to port some of the content from ```sphinx.sty``` into my template so that I can remove minted.

**Comment by mgeier** [comment]
Thanks @t-makaro!

My environment `nbsphinxfancyoutput` is really just a slight variation of `sphinxVerbatim` to get the same frames and breaking behavior, just without the source code formatting. It uses `framed.sty` to define framed boxes which allow page breaks.

I don't know if that's better or worse than `tcolorbox`, I'm just using whatever Sphinx uses.

> I'm curious how did you manage to make the prompts appear to the left of the cell?

No, there is no `minipage` involved! Initially I tried `\marginpar`, but that didn't work.

Then I tried using `\llap` from within the code cell itself. This is really nice because it generates text that extends to the left from the current point and just doesn't care if there is some other text or a page margin or whatever.

This works really well from within the `sphinxVerbatim` environment, because the prompt gets exactly the same font as the code and the vertical alignment with the first code line is perfect.

It's more problematic for the non-plain-text outputs (what I called "fancy"). Using `\llap{}` here produces an unwanted vertical shift in the contents. Using `\makebox[0pt][r]{}` worked a bit better, but the alignment of the prompt was off. I tried using `\smash[b]{}` on top of that (`\smash{}` didn't work well), which somewhat helped, but the alignment of the content and the prompt were still not perfect. I added a negative `\vskip` to the content which works fine for the moment.

But the alignment of the prompt is still not perfect. But what's worse, in some page breaks the prompt stays on the previous page and the cell output continues on the next page.

If you have ideas how to improve this, please tell me!

> You should test with h5 and h6 markdown headers directly before a code cell.  These get translated into `\paragraph` and `\subparagraph` in LaTeX. These don't line break before hand. It was causing me some issues.

This seems to work for me. What exactly was the problem there?

> What really interests me is line breaking.

I think that Sphinx somehow uses Pygments for this, but again, I'm just using what Sphinx does, I don't really know how it works ...

**Comment by t-makaro** [comment]
Ya, I originally tried both ```framed```, and ```mdframed```, for settling on ```tcolorbox``` as a simpler solution. But if you've made page breaking work with it, then they are basically the same.

Thanks for the tip with ```\llap```, I'll have to try that. I currently just use a negative ```\hspace``` and a negative ```\vspace``` that use lengths automatically set by the text in the prompt. I have the same issue with the prompt not following the code to the next page sometimes, but I did mitigate the issue by using a package called ```needspace```. It allows you to say that you need a minimum amount of space (like enough to place at least the first line of the code cell) before drawing the prompt. This only made the issue occur less frequently.

The problem that I had with h5/6 headers looked a little like:
![image](https://user-images.githubusercontent.com/18018386/46491388-75924000-c7bf-11e8-82c9-fd12c7e5ff7f.png)
instead of this:
![image](https://user-images.githubusercontent.com/18018386/46491418-89d63d00-c7bf-11e8-8d2d-67214adc97e1.png)

Both sphinx and nbconvert use Pygments for highlighting. In fact, so does latex's ```minted``` package. The only difference is that directly using Pygments for highlighting needs some manual work to make line breaking work. Minted also has the advantage that the theme for syntax highlighting can be changed at any time, so it doesn't need 1 theme for the whole document.

**Comment by mgeier** [comment]
OK, I didn't have that problem with h5 and h6 headers.

In the meantime, I got help from @jfbu for placing the prompts perfectly. See the latest progress on https://github.com/spatialaudio/nbsphinx/pull/218.
